File: app/src/main/java/org/exallium/tradetracker/app/controller/forms/TradeForm.java
package org.exallium.tradetracker.app.controller.forms;

import android.support.annotation.Nullable;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.EditText;
import butterknife.InjectView;
import com.google.common.collect.Iterables;
import com.orm.query.Condition;
import com.orm.query.Select;
import org.exallium.tradetracker.app.R;
import org.exallium.tradetracker.app.model.entities.Person;
import org.exallium.tradetracker.app.model.entities.Trade;
import org.exallium.tradetracker.app.utils.date.DateFormat;
import org.joda.time.LocalDate;
import rx.Observable;

public class TradeForm extends Form<Trade> {

    @InjectView(R.id.trade_person) AutoCompleteTextView tradePerson;
    @InjectView(R.id.trade_date) EditText tradeDate;

    private LocalDate cachedTradeDate = LocalDate.now();

    public TradeForm(Class<Trade> entityClass, View formView) {
        super(entityClass, formView);
        if (Select.from(Person.class).count() != 0) {
            Iterable<String> personNames = Observable.from(Select.from(Person.class).list()).map(person -> person.name).toBlocking().toIterable();
            String [] namesString = Iterables.toArray(personNames, String.class);
            tradePerson.setAdapter(new ArrayAdapter<>(formView.getContext(), R.layout.support_simple_spinner_dropdown_item, namesString));
        }
    }

    @Override
    public boolean isValid() {
        boolean valid = true;
        if (tradePerson.getText().length() == 0) {
            valid = false;
        }

        LocalDate newDate = DateFormat.fromString(tradeDate.getText().toString());
        if (newDate != null) {
            cachedTradeDate = newDate;
        } else {
            valid = false;
        }
        return valid;
    }

    @Override
    protected void populateEntity(@Nullable Trade entity) {
        if (entity != null && isValid()) {
            String name = tradePerson.getText().toString();
            Person person = Select.from(Person.class).where(Condition.prop("name").eq(name)).first();
            if (person == null) {
                person = new Person();
                person.name = name;
                person.save();
            }
            entity.person = person;
            entity.tradeDate = cachedTradeDate.toDate();
        } else if (entity != null) {
            entity.isTemporary = true;
            entity.tradeDate = cachedTradeDate.toDate();
        }
    }

    @Override
    protected void populateFields(@Nullable Trade entity) {
        if (entity != null) {
            cachedTradeDate = LocalDate.fromDateFields(entity.tradeDate);
            tradeDate.setText(DateFormat.toField(cachedTradeDate));
            tradePerson.setText(entity.person.name);
        }

    }
}

File: app/src/main/kotlin/org/exallium/tradetracker/app/controller/forms/TradeForm.kt
package org.exallium.tradetracker.app.controller.forms

import android.app.DatePickerDialog
import android.view.View
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.DatePicker
import android.widget.EditText
import butterknife.ButterKnifeViewHolder
import butterknife.bindView
import com.orm.query.Condition
import com.orm.query.Select
import org.exallium.tradetracker.app.R
import org.exallium.tradetracker.app.model.entities.Person
import org.exallium.tradetracker.app.model.entities.Trade
import org.exallium.tradetracker.app.utils.printForField
import org.exallium.tradetracker.app.utils.toLocalDate
import org.joda.time.LocalDate
import rx.android.view.ViewObservable
import java.util.Calendar

public class TradeForm(formView : View) : Form<Trade>(entityClass = javaClass<Trade>()) {

    class ViewHolder(view : View) : ButterKnifeViewHolder(itemView = view) {
        val tradePerson : AutoCompleteTextView by bindView(R.id.trade_person)
        val tradeDate : EditText by bindView(R.id.trade_date)
    }

    private val viewHolder = ViewHolder(formView)
    private var cachedDate : LocalDate = LocalDate.now()

    init {
        if (Select.from(javaClass<Person>()).count() != 0L) {
            val people = Select.from(javaClass<Person>()).list().map { person -> person?.name }
            viewHolder.tradePerson.setAdapter(ArrayAdapter<String>(viewHolder.tradePerson.getContext(), R.layout.support_simple_spinner_dropdown_item, people))
        }

        ViewObservable.clicks(viewHolder.tradeDate).subscribe {
            DatePickerDialog(viewHolder.tradeDate.getContext(), { datePicker: DatePicker?, year: Int, monthOfYear: Int, dayOfMonth: Int ->
                val calendar = Calendar.getInstance()
                calendar.set(Calendar.YEAR, year)
                calendar.set(Calendar.MONTH, monthOfYear)
                calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                cachedDate = LocalDate.fromCalendarFields(calendar)
                viewHolder.tradeDate.setText(cachedDate.printForField())
            }, cachedDate.getYear(), cachedDate.getMonthOfYear() - 1, cachedDate.getDayOfMonth()).show()
        }

    }

    override fun isValid(): Boolean {
        var valid = viewHolder.tradePerson.getText()?.length() != 0

        var date : LocalDate? = null
        try {
            date = viewHolder.tradeDate.getText()?.toString()?.toLocalDate()
        } catch (e : IllegalArgumentException) {
            // Invalid date
        }

        date?.let({
            cachedDate = date as LocalDate
        })

        valid = date != null && valid

        return valid
    }

    override fun populateFields(entity: Trade) {
        cachedDate = LocalDate.fromDateFields(entity.tradeDate)
        viewHolder.tradeDate.setText(cachedDate.printForField())
        viewHolder.tradePerson.setText(entity.person?.name)
    }

    override fun populateEntity(entity: Trade) {
        entity.isTemporary = !isValid()
        entity.person = getPerson()
        entity.tradeDate = cachedDate.toDate()
    }

    private fun getPerson(): Person? {
        var name = viewHolder.tradePerson.getText()?.toString()?:""
        if (name.length() == 0)
            return null

        var person = Select.from(javaClass<Person>()).where(Condition.prop("name").eq(name)).first()
        if (person == null) {
            person = Person()
            person.name = name
            person.save()
        }
        return person
    }


}
